hbase-server/src/main/java/org/apache/hadoop/hbase/coprocessor/RegionObserver.java

```diff
@@ -1104,4 +1105,18 @@ default DeleteTracker postInstantiateDeleteTracker(
       throws IOException {
     return delTracker;
   }
+
+  /**
+   * Called just before the WAL Entry is appended to the WAL. Implementing this hook allows
+   * coprocessors to add extended attributes to the WALKey that then get persisted to the
+   * WAL, and are available to replication endpoints to use in processing WAL Entries.
+   * @param ctx
+   * @param key
+   * @return
+   * @throws IOException
+   */
+  default WALKeyImpl preWALAppend(ObserverContext<RegionCoprocessorEnvironment> ctx, WALKeyImpl key)
```

Comment:
This only allows change of the WAL key. What if we want to also mutate or add more cells to the WALedit?

Reply:
Instead maybe `WALEntry preWALAppend(ctx, entry)` ?

Reply:
@apurtell - the header comments for WALEdit state that the class is meant to be read-only for coprocs. There are coprocessor hooks that take a WALEdit parameter, but unless I missed one, they all seem to be _after_ the WAL has been appended so that mutating the edit has no function. 

I don't understand the reason for the restriction -- we trust coprocessor implementers to do sane things already everywhere else -- but since the code had an explicit policy I tried to honor it.

Reply:
Also, it can't easily be WALEntry preWALAppend(ctx, entry), because doWALAppend does not currently create a WALEntry (it passes the WALKey and WALEdit as separate parameters)

Reply:
Ok, separate parameters is fine. Implementer can invoke WALKey and WALEdit methods to make changes.